### DISCUSSION TITLE
MySQL Upsert From Select

### DISCUSSION THREAD

**Discussion by mudgett06** [sqlalchemy/sqlalchemy#9523] [Usage Questions]
Hi, I am struggling to perform an upsert from a select clause. I am able to use the on_duplicate_key_update method to upsert values in bulk, but when using Insert.from_select(), the executed SQL throws a syntax error. Looking at the compiled SQL using mysql dialect, it differs from the compiled SQL using the engine. Here is what I have:

### To Reproduce
```python
from sqlalchemy import create_engine, select
from sqlalchemy.orm import Session, DeclarativeBase, Mapped, mapped_column
from sqlalchemy.dialects import mysql


class Base(DeclarativeBase):
    pass


class tbl1(Base):
    __tablename__ = 'table_1'
    col1: Mapped[int] = mapped_column(primary_key=True)
    col2: Mapped[int]


class tbl2(Base):
    __tablename__ = 'table_2'
    col1: Mapped[int] = mapped_column(primary_key=True)
    col2: Mapped[int]


stmt = mysql.insert(tbl1).from_select(tbl1.__table__.c, select(tbl2))
update_dict = {col.name: col for col in stmt.inserted}
upsert = stmt.on_duplicate_key_update(**update_dict)


print(str(upsert.compile(dialect=mysql.dialect())))
```
Prints
```
INSERT INTO table_1 (col1, col2) SELECT table_2.col1, table_2.col2 
FROM table_2 ON DUPLICATE KEY UPDATE col1 = VALUES(col1), col2 = VALUES(col2)
```
And
```python
engine = create_engine("mysql://user:password@host/db")
print(str(upsert.compile(engine)))
```
Prints
```
INSERT INTO table_1 (col1, col2) SELECT table_2.col1, table_2.col2 
FROM table_2 AS new ON DUPLICATE KEY UPDATE col1 = new.col1, col2 = new.col2
```
Note the aliasing of the select clause.
When Executed:

```python
with Session(engine) as session:
    session.execute(upsert)
```
I get this error:

```
(MySQLdb.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'AS new ON DUPLICATE KEY UPDATE col1= new.col1, date = new.col1, list_price = new.' at line 3")
```

I have no issues when running the text from the mysql.dialect() compiled query. Thanks in advance!

### Versions
OS: Ubuntu 22.04
Python: 3.10.6
SQLAlchemy: 2.0.6
MySQL Server: 8.0.32

**Comment by zzzeek** [comment]
hi - 

we have no test support for this pattern so we would need to know specifically what syntax is needed ( I cant see the problem just by looking at it) and we would then open an issue and put out a call for pull requests on this.

**Reply by mudgett06** [reply]
The syntax generated from `print(str(upsert.compile(dialect=mysql.dialect())))` is correct:

```sql
INSERT INTO table_1 (col1, col2) SELECT table_2.col1, table_2.col2 
FROM table_2 ON DUPLICATE KEY UPDATE col1 = VALUES(col1), col2 = VALUES(col2)
```

Is there any reason the compiled SQL would be different when using the engine vs. `mysql.dialect()`, givent that the engine's dialect is MySQL?

**Reply by zzzeek** [reply]
oh is the thing we're looking at this?

    VALUES(col1)

vs

    AS new .. new.col1

that's #8626 (https://docs.sqlalchemy.org/en/20/changelog/changelog_20.html#change-0d5c92a2cc411989e183f0e63afc3716) .  The second form, using an alias, is the correct one for MySQL 8.0 and above.  If your engine is using mariadb or an older MySQL version, you will get the VALUES version.

We don't emit VALUES(col1) on MySQL as of #8626 because MySQL emits a deprecation warning.

**Reply by zzzeek** [reply]
the difference between your engine vs. dialect is that your engine has connected to the DB and detected MySQL 8 in use.

**Reply by CaselIT** [reply]
~~the opposite appears to be happening. An uninitialized dialect is using the alias while the engine is using values. Is there a problem in the check?~~

sorry, I read it wrong.

**Reply by mudgett06** [reply]
Ah, I see. Today when I tried again, the syntax error went away, but the new error is this:

`(MySQLdb.OperationalError) (1054, "Unknown column 'table_2.col1' in 'field list'")
[SQL: INSERT INTO table_1 (col1, col2) SELECT table_2.col1, table_2.col2 
FROM table_2 AS new ON DUPLICATE KEY UPDATE col1 = new.col1, col2 = new.col2]`

It looks like this is because the original columns referenced from the selected table are not renamed to reference the aliased name "new"

**Reply by CaselIT** [reply]
that does seem like a bug

**Comment by POD666** [comment]
Is there any solution available? Looks like I have the same issue:
```py
from sqlalchemy.dialects.mysql import insert

insert_stmt = insert(cls.__table__).values(values)
on_duplicate_stmt = insert_stmt.on_duplicate_key_update(
    **{key: getattr(insert_stmt.inserted, key) for key in keys_to_update}
)
session.execute(on_duplicate_stmt)

...

sqlalchemy.exc.OperationalError: 
    (MySQLdb._exceptions.OperationalError) 
    (1054, "Unknown column 'new.target_meta_uid' in 'field list'")

[SQL: 
    INSERT INTO 
        task_run_metrics_v2 (task_run_attempt_id, source, name, value_string, timestamp) 
    VALUES 
        (%s, %s, %s, %s, %s) AS new 
    ON DUPLICATE KEY UPDATE 
        target_meta_uid = new.target_meta_uid, ...
]
```

**Reply by zzzeek** [reply]
I'm at the same point as I was before.  Someone show me the SQL they want which works

**Reply by POD666** [reply]
This works, pay attention to aliasing for values `... AS new(<fields>)`. Does it help?
```
INSERT INTO 
    task_run_metrics_v2 (task_run_attempt_id, source, name, value_string, timestamp) 
VALUES 
    (1, 'qwe', 'qwe', 'qwe', '2023-04-24') AS new(target_meta_uid, source, name, value_string, timestamp)
ON DUPLICATE KEY UPDATE 
    target_meta_uid = new.target_meta_uid, source = new.source, name = new.name, value_string = new.value_string, timestamp = new.timestamp
```
Tested on mysql:8.0

**Reply by zzzeek** [reply]
that's not at all the same thing that's asked for previously, that's a regular upsert with just one table, and we support that syntax right now (it's MySQL 8 specific). see #8626

**Reply by POD666** [reply]
Just checked the link, so I now I can point to the bug in code: https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/dialects/mysql/base.py#L1376-L1379

It adds an alias but without a column list like in my example. So it actually doesn't work.

It looks exactly like what was described in this comment, so I thought it was related https://github.com/sqlalchemy/sqlalchemy/discussions/9523#discussioncomment-5421912

Should I open a new issue instead?

**Reply by zzzeek** [reply]
that fix gives us this SQL:

```sql
            expected_sql = (
                "INSERT INTO foos (id, bar) VALUES (%s, %s), (%s, %s) "
                "AS new ON DUPLICATE KEY UPDATE "
                "bar = new.bar, "
                "baz = new.baz"
            )
```

and what you are requesting seems to be:


```sql
            expected_sql = (
                "INSERT INTO foos (id, bar) VALUES (%s, %s), (%s, %s) "
                "AS new(id, bar, baz) ON DUPLICATE KEY UPDATE "
                "bar = new.bar, "
                "baz = new.baz"
            )
```

the implication here would be that syntax number 1 is not working at all since it does not name the columns inside of `new()`, I can't see what else is different?   How does syntax 1 allow the phrase "new.bar" and "new.baz" if they are not named?

**Reply by POD666** [reply]
Exactly, syntax 1 (the current one) leads to errors like `(1054, "Unknown column 'new.target_meta_uid' in 'field list'")`.

So looks like columns inside of `new()` are required (syntax 2).

**Reply by zzzeek** [reply]
~~yes taht's a new issue.  the PR we received for this was wrong.~~

edit: this just seems to be an unnecessary naming change.   these are [just aliases](https://dev.mysql.com/doc/refman/8.0/en/insert-on-duplicate.html).  

below script should be equivalent, works on mysql 8


```py
import datetime

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import DateTime
from sqlalchemy import insert
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.dialects.mysql import insert
from sqlalchemy.orm import declarative_base

Base = declarative_base()


class TaskMetrics(Base):
    __tablename__ = "task_run_metrics_v2"

    task_run_attempt_id = Column(Integer, primary_key=True)
    target_metadata_uid = Column(Integer)
    source = Column(String(10))
    name = Column(String(10))
    value_string = Column(String(10))
    timestamp = Column(DateTime)


insert_stmt = insert(TaskMetrics.__table__).values(
    [
        {
            "task_run_attempt_id": 1,
            "source": "s1",
            "name": "n1",
            "value_string": "v1",
            "timestamp": datetime.datetime.now(),
        },
        {
            "task_run_attempt_id": 2,
            "source": "s2",
            "name": "n2",
            "value_string": "v2",
            "timestamp": datetime.datetime.now(),
        }
    ]
)
on_duplicate_stmt = insert_stmt.on_duplicate_key_update(
    target_metadata_uid=insert_stmt.inserted.task_run_attempt_id,
    source=insert_stmt.inserted.source,
)

e = create_engine("mysql://scott:tiger@mysql80/test", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)

with e.begin() as conn:
    conn.execute(on_duplicate_stmt)

```

**Reply by zzzeek** [reply]
OK inside of  `new()`, it cant be just random col names, they have to line up with the INSERT

in your SQL

```
INSERT INTO 
    task_run_metrics_v2 (task_run_attempt_id, source, name, value_string, timestamp) 
VALUES 
    (1, 'qwe', 'qwe', 'qwe', '2023-04-24') AS new(target_meta_uid, source, name, value_string, timestamp)
ON DUPLICATE KEY UPDATE 
    target_meta_uid = new.target_meta_uid, source = new.source, name = new.name, value_string = new.value_string, timestamp = new.timestamp

```

you are intentionally re-directing task_run_attempt_id to target_meta_uid.  

Why change the name?  how is your statement not equivalent to:


```
INSERT INTO 
    task_run_metrics_v2 (task_run_attempt_id, source, name, value_string, timestamp) 
VALUES 
    (1, 'qwe', 'qwe', 'qwe', '2023-04-24') AS new(task_run_attempt_id, source, name, value_string, timestamp)
ON DUPLICATE KEY UPDATE 
    target_meta_uid = new.task_run_attempt_id, source = new.source, name = new.name, value_string = new.value_string, timestamp = new.timestamp

```

?

these names in "new" are not table columns, they are just aliases to what you are sending in. they are not necessary if naming is abstracted away

**Reply by zzzeek** [reply]
https://dev.mysql.com/doc/refman/8.0/en/insert-on-duplicate.html

>  If, in addition, you use the column aliases m, n, and p, you can omit the row alias in the assignment clause and write the same statement like this:

> INSERT INTO t1 (a,b,c) VALUES (1,2,3),(4,5,6) AS new(m,n,p)
>   ON DUPLICATE KEY UPDATE c = m+n;


these are just aliases, not column names.  they are not needed

**Reply by POD666** [reply]
Ah, now I'm confused a bit. I didn't pay attention to it previously. 

I will check, perhaps there is some issue in my code.

Thank you!

**Reply by zzzeek** [reply]
thanks! sorry I dont follow these syntaxes either so i didnt know right off

**Reply by POD666** [reply]
Looks like keys_to_update in my python code was calculated wrong after sqla1.4 to sqla2 upgrade.

After fixing keys_to_update calculation all seems to work.

Thank you one more time!
